istioctl/cmd/internal-debug.go

```diff
@@ -139,57 +138,80 @@ By default it will use the default serviceAccount from (istio-system) namespace
 				TypeUrl: TypeDebug,
 			}
 
-			if centralOpts.Xds == "" {
-				svc, err := kubeClient.CoreV1().Services(istioNamespace).Get(context.Background(), istiodServiceName, metav1.GetOptions{})
-				if err != nil {
-					return fmt.Errorf("please specify %q as %v", "--xds-address", err)
-				}
-				namespace, selector, err := polymorphichelpers.SelectorsForObject(svc)
-				if err != nil {
-					return fmt.Errorf("please specify %q as we cannot attach to %T: %v", "--xds-address", svc, err)
+			svc, err := kubeClient.CoreV1().Services(istioNamespace).Get(context.Background(), istiodServiceName, metav1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("please specify %q as %v", "--xds-address", err)
+			}
+			podPort := 15012
+			for _, v := range svc.Spec.Ports {
+				if v.Name == xdsPortName {
+					podPort = v.TargetPort.IntValue()
 				}
+			}
 
-				options := metav1.ListOptions{LabelSelector: selector.String()}
+			labelSelector := centralOpts.XdsPodLabel
+			if labelSelector == "" {
+				labelSelector = "app=istiod"
+			}
+			podList, podErr := kubeClient.GetIstioPods(context.TODO(), istioNamespace, map[string]string{
+				"labelSelector": labelSelector,
+				"fieldSelector": "status.phase=Running",
+			})
+			if podErr != nil {
+				return podErr
+			}
+			if len(podList) == 0 {
+				return fmt.Errorf("no running Istio pods in %q", istioNamespace)
+			}
 
-				podList, err := kubeClient.CoreV1().Pods(namespace).List(context.TODO(), options)
-				if err != nil {
-					return fmt.Errorf("please specify %q as %v", "--xds-address", err)
-				}
-				//  select a pod randomly to simulate current debug behavior
-				pod := podList.Items[0]
-				podPort := 15012
-				for _, v := range svc.Spec.Ports {
-					if v.Name == xdsPortName {
-						podPort = v.TargetPort.IntValue()
+			var isNull bool = false
+			var printErr error
+			// Iterate all istiod pods for retrieving debug information
+			for _, pod := range podList {
+				fmt.Println("")
+				fmt.Println("------------------------------------------------------------------------------")
+				fmt.Println("--- istioctl x debug for pod: ", pod.Name)
```

Comment:
please add the namespace info as well.

Reply:
okay, it's very necessary to add the namespace of istiod pod. I will fix it and remove `---`. Thanks!